[Home page] Add Shop Products section with Products from the Catalog

adamraya opened the pull request:
<!--- Provide a short summary of your changes in the Title field above -->
GUS: https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000Q5jCCYAZ/view
# Description
This PR adds the Shop Products scroller to the Home page with Products from the first category on the catalog.

The products are obtained server-side using the `product-search` endpoint via the`getProps` API.

<!--- A longer summary of your changes, including: a description of the issue that you’re addressing, a list of required dependencies (if applicable), and any other relavant context. -->

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [X] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- Added Shop Products section using products from the `newarrivals` category.

# How to Test-Drive This PR

- Check out this branch `category-products`.
- Run `npm run start --prefix packages/pwa`
- Open the Network tab in Chrome Dev Tools.
- Load the Home page http://localhost:3000/
- Verify that the Shop Product section uses products from the first category in the catalog by sending a request to the `product-search` endpoint.

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

adamraya committed:
Create `useCategoryProducts()` hook and `<CategoryProducts />` component

adamraya committed:
Merge branch 'develop' into category-products

# Conflicts:
#	packages/pwa/app/pages/home/index.jsx

adamraya committed:
Cleanup

bendvc commented:
Hey @adamraya, I've take a look at your approach on this PR and I feel like instead of commenting on the code, I'll give you my feedback as a whole.

1. The HomePage is something that isn't ever going to be done in one way. What we decide isn't going to be what any of our clients are going to want to do. So for that reason I don't think it's wise to come up with generic components or hooks.

2. With this solution it doesn't looks like the content will be rendered on the server since we are using hooks to request the products used by the new CategoryProducts component. Is that intended? 


My opinion is that you should keep it simple. Use our API (getProps) and if the component isn't intended to be reused (the CategoryProducts component) don't make it a component. My rule of thumb is, if I'm making a component that is making requests to the backend on it's own, I need to have a really good reason to do that. Dependency injection (e.g. having a products property) is usually the easier correct solution. 

So here is the outline of my solution: 

1. I'd use **getProps** to get the products for a statically defined category id and limit (you choose this id, the catalog isn't going to change any time soon). 
_Note 1: If the data required to display in the product tiles isn't complete (e.g. not in the product search results) you can opt to make a secondary request or transform the data. (I might lean to doing the later)
Note 2: If we are intending to be lazy about loading this data (might be a good idea) you'll have to use an effect to get the data on the client-side only. But I would do this all in the home page._

2. Drop the new hook that you created since there is no expectation of use reusing that at the current time. Drop the use of the category context, we are oddly only using it to get a category id in a confusing way anyway. 

3. Same goes for the CategoryProduct component, I'd drop it and simply use the Product scroller directly in the page. But what I'm interested in is how much logic there is dedicated to adding this to wishlist, this seems like logic that is re-useable. Maybe talked to Kev about this since he's knowledgable about it. Maybe even ask if it's a requirement that we have those products be wishlist "addable" in the first place. If we have an example of how that works elsewhere, why repeat ourselves.

adamraya committed:
PR Feedback

adamraya commented:
Thank you for all the great feedback @bendvc.

1. 
> the products for a statically defined category id 

The App Generator asks "URL for your Commerce Cloud instance".

If users choose to use a custom ODS instance, they will probably have different categories. Thus, we can't assume a statically defined category id will exist in all the ODS instances. 
I chose to use one of the category id that exists in our sample ODS, 'newarrivals'. Users will not see the Shop Products section on the homepage if that category id doesn't exist in their custom ODS.

The previous solution proposed in this PR was to retrieve the first category id from the app context `categories` object containing all the existing categories in the ODS instance.

> use getProps to get the products

In the previous solution, I was using the hook approach client-side only to avoid creating an additional request server-side to obtain all the categories in `getProps` that we already store in the `categories` context client-side. But I agree that simpler is better. I implemented the logic in `getProps`.


2. 

Although I can see that is useful to get a list of products per category. I agree that `useCategoyProducts()` hook is not something that we reuse in the rest of the app. I implemented that logic into the homepage.

3. 

Agree. Now I'm using the `<ProductScroller />` component directly on the homepage. 
We're discussing via Slack if we also want to add the Wishlist handlers to the Shop Products section.

adamraya committed:
Merge branch 'develop' into category-products

vmarta commented on packages/pwa/app/pages/home/index.jsx:
```diff
@@ -191,17 +243,38 @@ const Home = () => {
         res.set('Cache-Control', 'max-age=31536000')
     }
+
+    const productSearchResult = await api.shopperSearch.productSearch({
+        parameters: {
+            refine: [`cgid=${SHOP_PRODUCTS_CATEGORY_ID}`, 'htype=master'],
+            limit: SHOP_PRODUCTS_LIMIT
+        }
+    })
```
For this PR, I agree that opting to hardcode this category id makes for a simpler homepage.

Topic for a future ticket/PR: 
But this brings to light the question of what to do if we want to fetch categories (e.g. to look up the id of the first category) in other pages' `getProps`. Like Adam said, we would want to avoid making duplicate calls to the categories endpoint. In other words, how do we deduplicate requests like that?

Perhaps we would need some logic in our Commerce API client that:
- caches duplicate requests
- manages the async calls ← in our scenario, App and the page-level component would attempt to fetch the categories at the same time

Perhaps instead of reinventing the wheel, we can explore using `react-query` library too.

bendvc replied:
I agree. This is probable a larger topic revolving around global data that we have purposefully avoided in the past. I think it's something we'll want to take a look at when the time is right and we aren't working on more important things. A natural time to look into this might be when we are looking at refactoring the commerce api hooks.

bendvc commented on packages/pwa/app/pages/home/index.jsx:
```diff
@@ -25,14 +25,18 @@ import Hero from '../../components/hero'
 import {heroFeatures, features} from './data'
+import ProductScroller from '../../components/product-scroller'
+
+const SHOP_PRODUCTS_CATEGORY_ID = 'newarrivals'
```
We have other values similar to this one for the PLP in the constants file located at `app/constants.js`. Although I don't see this constant ever being a configurable option in our new app config proposal, it might make more sense to be consistent (😅) with the placement

bendvc reviewed (approved):
PR looks a lot smaller and concise now. I left once small nit pick, but other than that it looks good.

adamraya committed:
PR Feedback

adamraya committed:
Merge branch 'develop' into category-products
